File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.graph = {}
        self.visited = set()
        self.ans = []

    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        # because tree cannot find parent, we need to build graph
        self.buildGraph(root, None)
        print(self.graph)
        self.visited = set()
        self.dfs(target.val, k)
        # if target.val in self.ans and k != 0:
        #     self.ans.remove(target.val)
        return self.ans
        
    def buildGraph(self, root: TreeNode, parent: TreeNode):
        if root is None:
            return
        if root.val not in self.visited:
            self.graph[root.val] = []
            self.visited.add(root.val)
        
            if root.left is not None:
                self.graph[root.val].append(root.left.val)
            if root.right is not None:
                self.graph[root.val].append(root.right.val)
            if parent is not None:
                self.graph[root.val].append(parent.val)

            self.buildGraph(root.left, root)
            self.buildGraph(root.right, root)
        
    
        
    def dfs(self, target: int, k: int):
        if k == 0:
            self.ans.append(target)
            return
        self.visited.add(target)
        for neighbor in self.graph[target]:
            if neighbor not in self.visited:
                self.visited.add(neighbor)
                self.dfs(neighbor, k - 1)
                    
        

File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// can traverse the tree and build a graph
// and find nodes with dist k
// space: O(E) then it's O(n) because E = n - 1
// time: traverse once and bfs: O(n)

// All the values Node.val are unique.

class Solution {
public:

    // implement a hashmap to store parent values
    unordered_map<int, TreeNode*> parents;

    void preorder(TreeNode* root, TreeNode* parentPtr) {

        if (!root)
            return;

        parents[root->val] = parentPtr;

        preorder(root->left, root);
        preorder(root->right, root);

    }

    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {

        preorder(root, nullptr);
        vector<int> ans;
        queue<TreeNode*> bfs;
        unordered_set<int> visited;
        bfs.push(target);
        visited.insert(target->val);
        while (!bfs.empty() && k >= 0) {
            int currLayerSize = bfs.size();
            while (currLayerSize) {

                TreeNode* curr = bfs.front();
                bfs.pop();

                cout << curr->val << endl;

                if (k == 0)
                    ans.push_back(curr->val);

                if (curr->left && visited.find(curr->left->val) == visited.end()) {
                    bfs.push(curr->left);
                    visited.insert(curr->left->val);
                }

                if (curr->right && visited.find(curr->right->val) == visited.end()) {
                    bfs.push(curr->right);
                    visited.insert(curr->right->val);
                }

                TreeNode* p = parents[curr->val];
                if (p && visited.find(p->val) == visited.end()) {
                    bfs.push(p);           
                    visited.insert(p->val);
                }
                currLayerSize--;
            }
            k--;
        }

        return ans;
        
    }
};